src/safeposix/vmmap.rs

```diff
@@ -0,0 +1,587 @@
+        }
+    }
+
+    fn find_page_iter_mut(
+        &mut self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)> {
+        if let Some(last_entry) = self.last_entry() {
+            self.entries
+                .overlapping_mut(ie(page_num, last_entry.0.end()))
+        } else {
+            // Return an empty iterator if no last_entry
+            self.entries.overlapping_mut(ie(page_num, page_num))
+        }
+    }
+
+    fn find_space(&self, npages: u32) -> Option<Interval<u32>> {
+        let start = self.first_entry();
+        let end = self.last_entry();
+
+        if start == None || end == None {
+            return None;
+        } else {
+            let start_unwrapped = start.unwrap().0.start();
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self
+                .entries
+                .gaps_trimmed(ie(start_unwrapped, end_unwrapped))
+            {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn find_space_above_hint(&self, npages: u32, hint: u32) -> Option<Interval<u32>> {
+        let start = hint;
+        let end = self.last_entry();
+
+        if end == None {
+            return None;
+        } else {
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self.entries.gaps_trimmed(ie(start, end_unwrapped)) {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn find_map_space(&self, num_pages: u32, pages_per_map: u32) -> Option<Interval<u32>> {
+        let start = self.first_entry();
+        let end = self.last_entry();
+
+        if start == None || end == None {
+            return None;
+        } else {
+            let start_unwrapped = start.unwrap().0.start();
+            let end_unwrapped = end.unwrap().0.end();
+
+            let rounded_num_pages =
+                self.round_page_num_up_to_map_multiple(num_pages, pages_per_map);
+
+            for gap in self
+                .entries
+                .gaps_trimmed(ie(start_unwrapped, end_unwrapped))
+            {
+                let aligned_start_page =
+                    self.trunc_page_num_down_to_map_multiple(gap.start(), pages_per_map);
+                let aligned_end_page =
+                    self.round_page_num_up_to_map_multiple(gap.end(), pages_per_map);
+
+                let gap_size = aligned_end_page - aligned_start_page;
+                if gap_size >= rounded_num_pages {
+                    return Some(ie(aligned_end_page - rounded_num_pages, aligned_end_page));
```

Comment:
why return the page from the end instead of the start? For sorting performance?

Reply:
may need to change this to return from start, need to figure out how to configure intervals